### PR TITLE
feat: style individual bars in a series

### PR DESCRIPTION
**Pull request by nickofthyme**
## Summary
Issue #216

Add ability to style(only color) individual bars in a series

Allow `colorAccessors` to be of type `AccessorFn`

https://github.com/elastic/elastic-charts/blob/79422c2d392e2912070b30d0efbbc36d58d3414e/src/lib/utils/accessor.ts#L3

### Checklist

- [ ] Any consumer-facing exports were added to `src/index.ts` (and stories only import from `../src` except for test data & storybook)
- [ ] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)
- [ ] Proper documentation or storybook story was added for features that require explanation or tutorials
- [ ] Unit tests were updated or added to match the most common scenarios
- [ ] Each commit follows the [convention](https://github.com/elastic/elastic-charts/blob/master/CONTRIBUTING.md)

### CODE REVIEW

**Comment by nickofthyme** [comment]
@markov00 & @emmacunningham This has proven to be a little trickier than I initially thought but has given me good insight into the lifecycle of a chart.

Currently, the `colorAccessor` prop changes the legend to add a new series but does _NOT_ affect the chart itself. After exploring the bar chart rendering I was able to get the chart to render the `colorAccessor` series in a separate series that bifurcates the entire `DataSeries`.

```ts
const colorAccessors = [(d) => d.y1 >= 2 ? 'nick' : null];
```

## Before
image

## After
image

I wanted to get both of your thoughts on this before I do anything else. From what I see all the rendering functions are assuming `DataSeries` has only one color. In order to achieve this feature, I am thinking to create a series that has "buckets" for each `colorSeriesKey`.

I am not sure of the overall implications for that sort of change. One thing that will be a bit problematic is the legend interaction with this new "nested" series.

Let me know your thoughts. Thanks!

> Note: I am just looking at the bar chart for now as I think there are design implications for the area/line chart that need to be discussed.

File: src/lib/utils/accessor.ts [hunk 1]
```diff
@@ -10,12 +10,18 @@ export type Accessor = AccessorString;
...
   throw new Error('Accessor must be a string or a function');
 }
+
+/**
+ * Returns if the accessor is an AccessorFn
+ * @param accessor the spec accessor
+ */
+export function isAccessorFn(accessor: Accessor | AccessorFn): boolean {
+  return typeof accessor === 'function';
+}
```
Comment (emmacunningham) [inline src/lib/utils/accessor.ts]: If you specify the return type as a [type predicate](https://www.typescriptlang.org/docs/handbook/advanced-types.html#type-guards-and-differentiating-types), then you won't need to do any explicitly casting to `AccessorFn` when the check passes (as is happening in `updateSeriesMap` where this function gets called).

File: src/lib/utils/domain.ts [hunk 2]
```diff
@@ -13,13 +13,6 @@ export interface SpecDomain {
   isStacked?: boolean;
 }
 
-export interface ColorDomain {
-  accessors: Accessor[];
-  yAccessors?: Accessor[];
-  domain: string[];
-  scaleType: ScaleType;
-}
-
```
Comment (emmacunningham) [inline src/lib/utils/domain.ts]: 👏

File: src/lib/series/series.ts [hunk 3]
```diff
@@ -169,25 +168,25 @@ function updateSeriesMap(
...
  */
-function getAccessorsValues(datum: Datum, accessors: Accessor[] = []): any[] {
+function getAccessorsValues(datum: Datum, accessors: Array<Accessor | AccessorFn> = []): any[] {
   return accessors
     .map((accessor) => {
-      return datum[accessor];
+      return isAccessorFn(accessor) ? (accessor as AccessorFn)(datum) : datum[accessor as Accessor];
     })
-    .filter((value) => value !== undefined);
+    .filter((value) => value !== undefined && value !== null);
```
Comment (emmacunningham) [inline src/lib/series/series.ts]: can be simplified to `value != null` since in javascript, `null == null` and `null == undefined`

**Comment by markov00** [comment]
@nickofthyme I think we forgot a bit about the `colorAccessors` prop. I think it was originally configured to do something near what we want to do here, but we probably didn't take care of that on most of the code refactoring :(
I also don't see any stories around `colorAccessors` so I think it's fair to change a bit it's destiny. @emmacunningham what do you think? 
The initial idea, and also how it's implemented seems that we want to to use the `colorAccessors` a way to "split" series by a different set of properties. I think the initial idea was something like:
I've 1 `<BarSeries>` with a dataset that looks like `[x, y, group, colorGroup]` and a configuration that split the series in 3 `splitAccessors={['x', 'group']}`. I want to color 2 series with the same color using a third parameter like `colorAccessors={[`colorGroup']}`. 
This will be ideallty translated in a 3 series bar chart with two colors and only 2 items on the legend. But, now that I'm thinking about it seems a bit strange.
I think we can refactor the way the `colorAccessor` is used in the way we really want to use: have the possibility to specify a color or a color index in the used color palette depending on the data series value. 
Few things to notice:
- I'm not completely sure that `colorAccessors` is the right name for that prop, we are specifying a function that change the color of a bar depending on the bar value, it will be worth having a look at other libraries and how they call similar functions/props(eg Victory from Formidable, vx or react-vis)  
- this function should be called with the value, with the `specId` and the data series key (usually identified by a key composed by the values coming from the splitAccessors). We may need also the automatically assigned color of the bar
- we should extend that function to lines and areas in the future, where we can have a line colored with a gradient between points, from color A to color B

**Comment by emmacunningham** [comment]
> I think we forgot a bit about the `colorAccessors` prop. I think it was originally configured to do something near what we want to do here, but we probably didn't take care of that on most of the code refactoring :(
> I also don't see any stories around `colorAccessors` so I think it's fair to change a bit it's destiny. @emmacunningham what do you think?

Yes, I agree!  (And I like the phrasing of "changing destiny" 😄 )

>     * I'm not completely sure that `colorAccessors` is the right name for that prop, we are specifying a function that change the color of a bar depending on the bar value, it will be worth having a look at other libraries and how they call similar functions/props(eg Victory from Formidable, vx or react-vis)

++ to thinking about a more accurate and descriptive name for this prop, especially now that we're altering destinies!

**Comment by nickofthyme** [comment]
Closed in favor of #255

**Pull request closed without merging**